### ISSUE TITLE
[cloud_firestore] Need to support local Cloud Firestore emulator

### ISSUE THREAD

**Issue by nonbeing** [FirebaseExtended/flutterfire#117]
**Is your feature request related to a problem? Please describe.**
There is support for Cloud Firestore emulator in other client libraries, for example in [javascript](https://github.com/firebase/quickstart-nodejs/tree/master/firestore-emulator/javascript-quickstart) and workarounds in other clients like [Python](https://stackoverflow.com/a/56386597/376240). To the best of my knowledge, there is no support (or at least, there isn't a document) in flutterfire yet, for the cloud firestore emulator, even if just for testing out Firestore Security Rules locally in my devenv.

**Describe the solution you'd like**
I would like a document explaining how to connect to and use the local Cloud Firestore emulator that can be started easily like this:

```
> gcloud beta emulators firestore start
Executing: cmd /c C:\Users\ambar\AppData\Local\Google\Cloud SDK\google-cloud-sdk\platform\cloud-firestore-emulator\cloud_firestore_emulator.cmd start --host=::1 --port=8372
[firestore] API endpoint: http://::1:8372
[firestore] If you are using a library that supports the FIRESTORE_EMULATOR_HOST environment variable, run:
[firestore]
[firestore]    export FIRESTORE_EMULATOR_HOST=::1:8372
[firestore]
[firestore] Dev App Server is now running.
[firestore]
```
I would be grateful if there were a simple guide to explain how to connect to the emulator from Flutter, set local Firestore security rules, and then test this out in a basic Flutter app. Perhaps something similar to this quickstart guide: https://github.com/firebase/quickstart-nodejs/tree/master/firestore-emulator/javascript-quickstart

**Additional context**
Why am I asking for this? I am running into problems trying to get my flutter app to work my GCP Cloud Firestore instance, probably because of slightly complex FS security rules. There are no server-side logs and no client-side hints why my requests are failing, so I decided to try the local Firestore emulator for troubleshooting this. But I don't know how to connect to the local FS emulator from Flutter.

Please correct me if I'm wrong or missing something, and point me to the relevant doc, if there is one.

**Comment by Levi-Lesches** [comment]
Connecting to the emulator would imply a dev environment only solution, there would have to be different systems in place for dev and release **in your codebase**. I don't work for flutter by the way, so maybe this could still work, but I'd warn against it.

On the other hand, if you have a specific question about Firebase that's stopping you from using it (or help on how to use the emulators locally), we'd be glad to help with more specific details.

**Comment by nonbeing** [comment]
Yes, I need help connecting to the Firestore emulator from Flutterfire. Would be grateful for instructions on how to do this.

**Comment by Levi-Lesches** [comment]
> Connecting to the emulator would imply a dev environment only solution

Again, it's not a good idea to connect to the emulator, since you cannot do this in your app, which is what FlutterFire is designed for.

**Comment by nonbeing** [comment]
Like I said in my first post, this is required only for my dev environment. My app is still in development, and I am not envisioning going to production any time soon. I am unable to troubleshoot a problem with Firestore security rules with Cloud Firestore (lack of logs being one reason). I need to debug the problem on the Firestore emulator with a copy of the same security rules I am using on the Cloud. This is one of the primary use-cases for having a local (i.e. dev-environment) Firestore emulator, as explained by the Firestore team too - the other major use-case is running unit tests against the emulator in a CI/CD fashion.

Once I resolve the problem locally, in my dev environment, then I will use the same Firestore security rules for my Cloud Firestore in GCP.

I hope I have been able to justify my need for using the emulator, and I hope you appreciate this now. Would be grateful if you could share any guide/instructions for using the emulator.

**Comment by Levi-Lesches** [comment]
Oh yeah I agree security rules not having logs are really terrible. But I was able to solve my problem by running Node.js unit tests against the emulator. More details can be found on the [official Firebase docs](https://firebase.google.com/docs/firestore/security/test-rules-emulator?authuser=0). After running the tests, you will get a `localhost` URL that will show you interactively how exactly your rules failed. This and the fact that you can now unit test makes using Javascript worth it. 

That being said I acknowledge it would be much easier to use the emulator in dev mode, but Flutter is about keeping your codebase the same, so I don't think this will be introduced. But again, I don't work with Flutter or Firebase, so take my opinion with a grain of salt.

**Comment by Salakar** [comment]
Hey, as part of our on-going work for #2582, this has been resolved in our Firebase Firestore rework (#2913) (Firestore instance settings can be set to use the emulator work with Hot Restarts (no need to fully quit the application anymore) - which has now been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
